.github/workflows/dockerhub_ci.yml

```diff
@@ -21,115 +21,125 @@ jobs:
-            fi
-            
-            BUILD_ARGS="OPENJDK_VERSION=${OPENJDK_VERSION}
-                        HADOOP_VERSION=${HADOOP_VERSION}
-                        SPARK_VERSION=${DOCKER_TAG%%,*}"
-
-            # No modification needed beyond this point
-            BUILD_ARGS="${BUILD_ARGS//'%'/'%25'}"
-            BUILD_ARGS="${BUILD_ARGS//$'\n'/'%0A'}"
-            BUILD_ARGS="${BUILD_ARGS//$'\r'/'%0D'}"
-
-            echo "::set-output name=build_args::$BUILD_ARGS"
-
-      - name: Prepare
-        id: prep
-        run: |
-            DOCKER_TAG="${{ matrix.DOCKER_TAG }}"
-            DOCKER_IMAGE_NAME="$DOCKER_REPO:${DOCKER_TAG%%,*}"
-            echo ::set-output name=image_name::${DOCKER_IMAGE_NAME}
-            
-            TAGS="$DOCKER_REPO:${DOCKER_TAG//,/,$DOCKER_REPO:}"
-            echo ::set-output name=tags::${TAGS}
-
-            echo ::set-output name=created::$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+        uses: actions/checkout@v4
 
       - name: Set up QEMU
-        uses: docker/setup-qemu-action@v2
+        uses: docker/setup-qemu-action@v3
 
       - name: Set up Docker Buildx
-        uses: docker/setup-buildx-action@v2
+        id: setup-buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Generate image metadata
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: "${{ env.DOCKER_REPO }}"
+          flavor: |
+            latest=${{ matrix.image.LATEST }}
+            prefix=
+            suffix=
+          tags: |
+            type=semver,pattern={{version}},value=${{ matrix.image.SPARK_VERSION }}
+            type=semver,pattern={{major}}.{{minor}},value=${{ matrix.image.SPARK_VERSION }}
+          labels: |
+            org.opencontainers.image.source=${{ github.event.repository.clone_url }}
+            org.opencontainers.image.revision=${{ github.sha }}
+            org.opencontainers.image.vendor=Xebia Data, https://xebia.com/
+            org.opencontainers.image.description=An Apache Spark Docker image.
 
       - name: Build image
-        uses: docker/build-push-action@v4
+        id: docker-build
+        uses: docker/build-push-action@v6
         with:
          context: .
          file: ./Dockerfile
          push: false
          load: true
-         tags: ${{ steps.prep.outputs.tags }}
+         tags: ${{ steps.meta.outputs.tags }}
          build-args: |
-           ${{ steps.args.outputs.build_args }}
-         labels: |
-           org.opencontainers.image.source=${{ github.event.repository.clone_url }}
-           org.opencontainers.image.created=${{ steps.prep.outputs.created }}
-           org.opencontainers.image.revision=${{ github.sha }}
+           HADOOP_VERSION=${{ matrix.image.HADOOP_VERSION }}
+           OPENJDK_VERSION=${{ matrix.image.OPENJDK_VERSION }}
+           SPARK_VERSION=${{ matrix.image.SPARK_VERSION }}
+         labels: ${{ steps.meta.outputs.labels }}
+         annotations: ${{ steps.meta.outputs.annotations }}
```

Comment:
In the meta action we set labels. How does that relate to annotations?

Reply:
Ik snapte alleen niet waar de annotaties vandaan komen aangezien we ze zelf niet zetten in de meta step toch?

Reply:
Alles wat er staat is afkomstig uit de GitHub context (commit ID enzo) en de `tags:` data. De vendor en description ontbreken; die had ik zelf toegevoegd aan labels.

Reply:
Het zou best kunnen dat deze twee regels in de CI yaml helemaal niet nodig zijn omdat ze een default hebben:

```diff
          labels: |
-            org.opencontainers.image.source=${{ github.event.repository.clone_url }}
-            org.opencontainers.image.revision=${{ github.sha }}
            org.opencontainers.image.vendor=Xebia Data, https://xebia.com/
            org.opencontainers.image.description=An Apache Spark Docker image.
```

Reply:
https://github.com/docker/metadata-action/blob/b53be03109c4ef6f6cc7aa545b84b17a7fe51c1e/src/meta.ts#L509-L519 bingo. Als ik die description kopieer naar de repo description op de frontpagina kan ik hem volgens mij verwijderen uit de CI file

Reply:
Labels and annotations now match:

```json
Bake file definition (labels)
  {
    "target": {
      "docker-metadata-action": {
        "labels": {
          "org.opencontainers.image.created": "2024-11-29T09:38:33.748Z",
          "org.opencontainers.image.description": "An Apache Spark Docker image.",
          "org.opencontainers.image.licenses": "Apache-2.0",
          "org.opencontainers.image.revision": "b2e1a9f838eeaec28ef02ee37214691096b56c7b",
          "org.opencontainers.image.source": "https://github.com/godatadriven-dockerhub/spark",
          "org.opencontainers.image.title": "spark",
          "org.opencontainers.image.url": "https://github.com/godatadriven-dockerhub/spark",
          "org.opencontainers.image.vendor": "Xebia Data, https://xebia.com/",
          "org.opencontainers.image.version": "3.5.3"
        }
      }
    }
  }
Bake file definition (annotations)
  {
    "target": {
      "docker-metadata-action": {
        "annotations": [
          "manifest:org.opencontainers.image.created=2024-11-29T09:38:33.748Z",
          "manifest:org.opencontainers.image.description=An Apache Spark Docker image.",
          "manifest:org.opencontainers.image.licenses=Apache-2.0",
          "manifest:org.opencontainers.image.revision=b2e1a9f838eeaec28ef02ee37214691096b56c7b",
          "manifest:org.opencontainers.image.source=https://github.com/godatadriven-dockerhub/spark",
          "manifest:org.opencontainers.image.title=spark",
          "manifest:org.opencontainers.image.url=https://github.com/godatadriven-dockerhub/spark",
          "manifest:org.opencontainers.image.vendor=Xebia Data, https://xebia.com/",
          "manifest:org.opencontainers.image.version=3.5.3"
        ]
      }
    }
  }
```